src/EditorFeatures/Core/Shared/Preview/PreviewSolutionCrawlerRegistrationService.cs

```diff
@@ -68,6 +74,34 @@ public void Unregister(Workspace workspace, bool blockingShutdown = false)
             {
                 source.Cancel();
             }
+
+            ClearDiagnostics(workspace);
+
+            AssertCleanup();
+        }
+
+        private void ClearDiagnostics(Workspace workspace)
+        {
+            var analyzer = _provider.CreateIncrementalAnalyzer(workspace);
+            var args = _service.GetDiagnosticsUpdatedEventArgs(workspace, projectId: null, documentId: null, cancellationToken: CancellationToken.None);
+
+            foreach (var documentId in args.Select(a => a.DocumentId).Distinct())
+            {
+                analyzer.RemoveDocument(documentId);
```

Comment:
> it has DocumentResetAsync but unfortunately I can't use it since it is async method. for now, I need to stick with RemoveDocument.

Note: because you are clearing the data associate for the workspace, the methods you call should not be document-centric.  They should be workspace-centric.  Otherwise we have the wrong level of granularity, and it's totally possible for an analyzer to be caching data that won't get removed.

--

Note: this is not a hypothetical.  This is how one of our incremental analyzers works today. SymbolTreeIncrementalAnalyzer has this code:

```c#
            public override Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, CancellationToken cancellationToken)
            {
                if (!document.SupportsSyntaxTree)
                {
                    // Not a language we can produce indices for (i.e. TypeScript).  Bail immediately.
                    return SpecializedTasks.EmptyTask;
                }

                if (bodyOpt != null)
                {
                    // This was a method level edit.  This can't change the symbol tree info
                    // for this project.  Bail immediately.
                    return SpecializedTasks.EmptyTask;
                }

                return UpdateSymbolTreeInfoAsync(document.Project, cancellationToken);
            }
```

Note that at hte bottom, we do our processing on the *Project* of the document and we cache that data at the project level.

"RemoveDocument" isn't even implemented by this class, so no cleanup happens at all.

--

Now, i have no idea if this is a problem or not.  But it seems *really* fishy that we're just calling "RemoveDocument" and we're expecting that to be sufficient.